text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,441 @@
+
+This part is a two element array with the crate name and the JSON representation of a type implementation.
+
+* key: `trait-impl`, for `doc/trait.impl/**/*.js`
+
+This part is a two element array with the crate name and the JSON representation of a trait implementation.
+
+## New flag: `--write-info-json=path/to/crate-info.json`
+
+When this flag is provided, the unmerged parts for the current crate will be written to `path/to/crate-info.json`. A typical `<path to crate-info.json>` is `./target/doc.parts/<crate name>/crate-info.json`.
+
+Crates `--include-info-json`ed will not appear in `crate-info.json`, as `crate-info.json` only includes the CCI parts for the current crate.
+
+If this flag is not provided, no `crate-info.json` will be written.
+
+## New flag: `--include-info-json=path/to/crate-info.json`
+
+If this flag is provided, rustdoc will expect `path/to/crate-info.json` to be the `crate-info.json` file containing the parts for a crate. It will append these parts to the ones it will render in the doc root (`--out-dir`). The info that's included is not written to `crate-info.json`, as `crate-info.json` only holds the CCI parts for the current crate.
+
+This flag is similar to `--extern-html-root-url` in that it only needs to be provided for externally documented crates. The flag `--extern-html-root-url` controls hyperlink generation. The hyperlink provided in `--extern-html-root-url` never accessed by rustdoc, and represents the final destination of the documentation. The new flag `--include-info-json` tells rustdoc where to search for the `crate-info.json` directory at documentation-time. It must not be a URL.
+
+In the Guide-level explanation, for example, crate `i` needs to identify the location of `s`'s parts. Since they could be located in an arbitrary directory, `i` must be instructed on where to fetch them. In this example, `s`'s parts happen to be in `./s/target/doc.parts/s`, so rustdoc is called with `--include-info-json=s/target/doc.parts/s/crate-info.json`.
+
+## New flag: `--merge=auto|none|write-only`
+
+This flag controls two internal paramaters: `read_rendered_cci`, and `write_rendered_cci`.
+
+When `write_rendered_cci` is active, rustdoc will output the rendered parts to the doc root (`--out-dir`). Rustdoc will generate files like `doc/search-index.js`, `doc/search.desc`, `doc/index.html`, etc if and only if this parameter is true.
+
+When `read_rendered_cci` is active, rustdoc will look in the `--out-dir` for rendered cross-crate info files. These files will be used as the base. Any new parts that rustdoc generates with its current invocation and any parts fetched with `include-info-json` will be appended to these base files. When it is disabled, the cross-crate info files start empty and are populated with the current crate's info and any crates fetched with `--include-info-json`.
+
+* `--merge=auto` (`read_rendered_cci && write_rendered_cci`) is the default, and reflects the current behavior of rustdoc. 
```

Comment:
I'm not a big fan of `auto`. Why not `read-write` instead? Would be more explicit (took me a while to get why it was different to `write-only` :sweat_smile:).

Reply:
That sounds good!